tetris_3.py

```diff
@@ -75,6 +54,40 @@ def tetris(d):
     return string
 
 
+def update_board_with_piece(array, entry):
+    piece_array = PIECES[entry['piece']]
+    column = entry['column']
+    al = len(piece_array)  # array length
+    # takes part of a piece individually, not good for Z and S
+    while al >= 1:
+        part = piece_array[al - 1]  # start at the last/bottom part
+        for string in part:
+            pl = len(string)
+        row = 0
+        while row < 21:
+            if all(array[row * 13 + column + space] is ws
+                   for space in range(0, pl)):
+                # if all spaces below the piece are blank, return true
+                row += 1
+            else:
+                array[(row-1)*13 + column] = part
```

Comment:
I think you've got a problem here. you originally defined `array` as an array (by the way, probably need a better name for that variable, maybe `board`) of 1 char long strings, and here you're assigning a whole piece row (`part`) which potentially is multiple chars long, to a single position in `array`. you end up with potentially a section of the piece in kind of a 'ghost' location and that's why you have to do the `dl` trick in the following lines. you might as well assign the piece chars directly to their right location on the board (`array`)